hivemind/averaging/allreduce.py

```diff
@@ -240,62 +301,62 @@ async def _accumulate_parts():
                     await accumulate_task
 
             except Exception as e:
-                self.finalize(exception=e)
+                logger.warning(f"Caught {repr(e)} when communicating with {context.remote_id}")
+                async with self.banlock:
+                    if context.remote_id not in self.banned_senders:
+                        self.banned_senders.add(context.remote_id)
+                        self.tensor_part_reducer.on_sender_failed(sender_index)
                 yield averaging_pb2.AveragingData(code=averaging_pb2.INTERNAL_ERROR)
         else:
             error_code = averaging_pb2.MessageCode.Name(request.code)
-            logger.debug(f"{self} - peer {context.remote_id} sent {error_code}, allreduce cannot continue")
-            self.finalize(exception=AllreduceException(f"Peer {context.remote_id} sent {error_code}"))
+            logger.debug(f"{self} - peer {context.remote_id} sent {error_code}")
+            sender_index = self.sender_peer_ids.index(context.remote_id)
+            async with self.banlock:
+                if context.remote_id not in self.banned_senders:
+                    self.banned_senders.add(context.remote_id)
+                    self.tensor_part_reducer.on_sender_failed(sender_index)
```

Comment:
Please fix this code duplication (L305-308, L314-317, L352-355), this is likely to lead to a bug if someone decides to change the ban procedure (but changes only 1-2 of 3 copies).

Reply:
replaced with a unified `await self._ban_sender(peer_id)` call